src/components/tests/SourcesTree.spec.js

```diff
@@ -6,32 +6,617 @@ import React from "react";
 import { shallow } from "enzyme";
 import SourcesTree from "../../components/PrimaryPanes/SourcesTree";
 import * as I from "immutable";
+import { showMenu } from "devtools-contextmenu";
+import { copyToTheClipboard } from "../../utils/clipboard";
+import {
+  getShownSource,
+  getSelectedSource,
+  getDebuggeeUrl,
+  getExpandedState,
+  getProjectDirectoryRoot,
+  getSources
+} from "../../selectors";
+
+jest.mock("devtools-contextmenu", () => ({ showMenu: jest.fn() }));
+jest.mock("../../utils/clipboard", () => ({ copyToTheClipboard: jest.fn() }));
+jest.mock("../../selectors");
+
+const defaultSources = I.Map({
+  "server1.conn13.child1/39": createMockSource(
+    "server1.conn13.child1/39",
+    "http://mdn.com/one.js"
+  ),
+  "server1.conn13.child1/40": createMockSource(
+    "server1.conn13.child1/40",
+    "http://mdn.com/two.js"
+  ),
+  "server1.conn13.child1/41": createMockSource(
+    "server1.conn13.child1/41",
+    "http://mdn.com/three.js"
+  )
+});
+
+const singleMockSource = I.Map({
+  "server1.conn13.child1/41": createMockSource(
+    "server1.conn13.child1/41",
+    "http://mdn.com/three.js"
+  )
+});
+
+const singleMockItem = createMockItem(
+  "http://mdn.com/one.js",
+  "one.js",
+  I.Map({ id: "server1.conn13.child1/39" })
+);
+
+const blackboxedMockItem = createMockItem(
+  "http://mdn.com/blackboxed.js",
+  "blackboxed.js",
+  I.Map({ id: "server1.conn13.child1/59" })
+);
+
+const blackboxedMockSource = I.Map({
+  "server1.conn13.child1/59": createMockSource(
+    "server1.conn13.child1/59",
+    "http://mdn.com/blackboxed.js",
+    true
+  )
+});
+
+const mockRootDirectory = {
+  contents: [],
+  name: "root",
+  path: "root/"
+};
+
+const mockDirectory = {
+  contents: [],
+  name: "folder",
+  path: "folder/"
+};
+
+const mockEvent = {
+  preventDefault: jest.fn(),
+  stopPropagation: jest.fn()
+};
+
+let defaultComponent;
+let defaultProps;
+let defaultState;
+let componentWithRoot;
+let propsWithRoot;
+let emptyComponent;
+let componentWithMdnRoot;
+let propsWithMdnRoot;
+let blackboxedComponent;
+
+describe("SourcesTree", () => {
+  beforeEach(() => {
+    let rendered = render();
+    defaultComponent = rendered.component;
+    defaultProps = rendered.props;
+    defaultState = defaultComponent.state();
+
+    rendered = render({
+      projectRoot: "root/"
+    });
+    componentWithRoot = rendered.component;
+    propsWithRoot = rendered.props;
+
+    emptyComponent = render({
+      projectRoot: "custom",
+      sources: I.Map()
+    }).component;
+
+    rendered = render({
+      projectRoot: "mdn"
+    });
+    componentWithMdnRoot = rendered.component;
```

Comment:
regarding lines 100-114. You do not want to reassign variables in general. In this case it is very confusing, because I have to really pay attention to which `rendered` object has which effect. it makes the code really hard to follow. One trick to avoid doing this kind of thing is -- imagine the next person who reads this code is drunk -- would they be able to tell you what it is doing? 

again, for this kind of setup -- put the variables in the test, not in the before each or in globals. Avoid globals unless they do not change.  

change means: reassignment or state change related to what you are testing. Something like the mocks you are using is fine. However: creating a lot of test state related variables in the globals scope will lead to problems.